src/foam/dao/pg/PostgresDAO.java

```diff
@@ -0,0 +1,326 @@
+      builder.append(" values");
+      buildFormattedColumnPlaceholders(obj, builder);
+      builder.append(" on conflict (id) do update set");
+      buildFormattedColumnNames(obj, builder);
+      builder.append(" = ");
+      buildFormattedColumnPlaceholders(obj, builder);
+
+      int index = 1;
+      PreparedStatement stmt = c.prepareStatement(builder.toString(),
+          Statement.RETURN_GENERATED_KEYS);
+      // set statement values twice: once for the insert and once for the update on conflict
+      index = setStatementValues(index, stmt, obj);
+      index = setStatementValues(index, stmt, obj);
+
+      int inserted = stmt.executeUpdate();
+      if (inserted == 0) {
+        throw new SQLException("Error performing put_ command");
+      }
+
+      // get auto-generated postgres keys
+      ResultSet resultSet = stmt.getGeneratedKeys();
+      if ( resultSet.next() ) {
+        obj.setProperty("id", resultSet.getLong(1));
+      }
+
+      resultSet.close();
+      stmt.close();
+      c.close();
+    } catch (SQLException e) {
+      e.printStackTrace();
+    }
+
+    return obj;
+  }
+
+  /**
+   * Creates a table if one does not exist already
+   * @throws SQLException
+   */
+  public void createTable(ClassInfo classInfo) throws SQLException {
+    Connection conn = connectionPool.getConnection();
+    DatabaseMetaData meta = conn.getMetaData();
+    ResultSet resultSet = meta.getTables(null, null, table, new String[] { "TABLE" });
+    if ( resultSet.isBeforeFirst() ) {
+      // found a table, don't create
+      return;
+    }
+
+    List<PropertyInfo> props = classInfo.getAxiomsByClass(PropertyInfo.class);
+    String columns = props.stream()
+        .filter(e -> ! "id".equals(e.getName()) && ! e.getStorageTransient())
+        .map(e -> {
+          // map types to postgres types
+          SQLType type = e.getSqlType();
+          switch ( type.getName() ) {
+            case "TINYINT":
+              return e.getName() + " SMALLINT";
+            case "VARBINARY":
+              return e.getName() + " BYTEA";
+            default:
+              return e.getName() + " " + type.getName();
+          }
+        })
+        .collect(Collectors.joining(","));
+
+    StringBuilder builder = sb.get()
+        .append("CREATE TABLE ")
+        .append(table)
+        .append("(id bigserial primary key,")
+        .append(columns)
+        .append(")");
+
+    // execute statement
+    PreparedStatement stmt = conn.prepareStatement(builder.toString());
+    stmt.executeUpdate();
+
+    resultSet.close();
+    stmt.close();
+    conn.close();
+  }
+
+  /**
+   * maps a database result row to an FObject
+   *
+   * @param resultSet
+   * @return FObject
+   * @throws SQLException
+   */
+  private FObject createFObject(ResultSet resultSet) throws Exception {
+    if ( getOf() == null ) {
+      throw new Exception("`Of` is not set");
+    }
+
+    List<PropertyInfo> props = getOf().getAxiomsByClass(PropertyInfo.class);
```

Comment:
You should cache a list of non-transient properties as an instance variable to avoid creating list every time.